src/components/calcite-popover/calcite-popover.tsx

```diff
@@ -0,0 +1,217 @@
+
+  @Element() el: HTMLCalcitePopoverElement;
+
+  @State() popper: Popper;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Lifecycle
+  //
+  // --------------------------------------------------------------------------
+
+  componentDidLoad() {
+    this.reposition();
+  }
+
+  componentDidUnload() {
+    this.destroyPopper();
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Public Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method() async reposition(): Promise<void> {
+    const {
+      el,
+      placement: componentPlacement,
+      open,
+      popper,
+      referenceElement,
+      xOffset,
+      yOffset
+    } = this;
+
+    const placement =
+      componentPlacement === "vertical" ? "bottom-start" : "auto-start";
+    const offsetEnabled = !!(yOffset || xOffset);
+
+    const modifiers: Popper.Modifiers = {
+      hide: {
+        enabled: false
+      },
+      offset: {
+        enabled: offsetEnabled,
+        offset: `${yOffset}, ${xOffset}`
+      },
+      preventOverflow: {
+        enabled: false
+      }
+    };
+
+    if (popper) {
+      popper.options.placement = placement;
+      popper.options.modifiers = { ...popper.options.modifiers, ...modifiers };
+      popper.scheduleUpdate();
+      return;
+    }
+
+    if (referenceElement && open) {
+      const newPopper = new Popper(referenceElement, el, {
+        eventsEnabled: false,
+        placement,
+        modifiers,
+        onCreate: data => {
+          if (
+            data.originalPlacement === "bottom-start" &&
+            document.body.clientWidth &&
+            data.offsets &&
+            data.offsets.reference &&
+            data.offsets.reference.left > document.body.clientWidth / 2
+          ) {
+            data.instance.options.placement = "bottom-end";
+            data.instance.scheduleUpdate();
+          }
+        }
+      });
+
+      window.addEventListener("resize", newPopper.scheduleUpdate, {
+        passive: true
+      });
+
+      this.popper = newPopper;
+    }
+  }
+
+  @Method() async toggle(): Promise<void> {
+    this.open = !this.open;
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  // --------------------------------------------------------------------------
+
+  destroyPopper(): void {
+    const { popper } = this;
+
+    if (popper) {
+      window.removeEventListener("resize", popper.scheduleUpdate);
+      popper.destroy();
+    }
+
+    this.popper = null;
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Render Methods
+  //
+  // --------------------------------------------------------------------------
+
+  render() {
+    return (
+      <Host>
+        <div
+          class={classnames(CSS.container, { [CSS.containerOpen]: this.open })}
```

Comment:
IMO this would be clearer without the third party dependency/resources object:

```
class={{
  "container": true,
  "container--open": this.open
}}
```

Reply:
Also, BEM would probably be something more like `calcite-popover calcite-popover--open`

Reply:
> Also, BEM would probably be something more like calcite-popover calcite-popover--open

Yeah. We've been using "partial-BEM" since we really only need the modifier because it will be in the shadowDOM so there won't be any conflicts outside of itself anyways. I can change if you want. Let me know

Reply:
Yeah, I don't super care, not a deal breaker for me for sure. There is definitely an argument to be made that shorter class names are more performant. We could theoretically go back and update all our existing components to use this strategy as well.

Reply:
I'd suggest just having "open" or "isOpen" class. With the shadow dom naming collisions are a thing of the past :D

Reply:
I don't think they documented it... 🤦‍♂ 

I found out about it reading through their 1.0 breaking changes: https://github.com/ionic-team/stencil/blob/master/BREAKING_CHANGES.md#hostdata

Reply:
😆 in true Stencil fashion. I've found several hidden features this way as well.